rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,106 @@
+helpful to the user to be able to specify if they prefer to use a kernel that
+has been designed to be reusable via the kernel cache (accepting any possible
+performance impacts associated with it) or the typical performance-optimized
+kernel. This proposal is to add a knob the user can control to specify whether
+they would like reusable kernels to be dispatched preferentially over the
+standard performance-optimized kernel ordering for GPU implementations.
+
+> **Note:** Reusable kernels are defined as GPU kernels with a finite set of
+configurations, which have opted-into the kernel cache internal API. Reusable
+kernels may depend on problem shape, but only in a limited way (i.e. the
+specific shape cannot be compiled directly into the kernel).
+
+## Proposal
+
+The proposal has four potential options: using DNNL_RUNTIME_DIM_VAL, a build
+flag, an environment variable, or a primitive attribute. In any of the options,
+the default behaviour will be the current one: kernels are ordered according to
+internal heuristics, roughly in order of decreasing likelihood for optimal
+performance for the given problem.
+
+If the reusable kernel option is selected, GPU kernels designed to be reusable
+will be selected first. In the case that no reusable kernels are compatible with
+the problem, kernel selection will continue with non-reusable kernels.
+
+Option 4 (primitive attribute) is the preferred option by frameworks, as they
+don't know which dimensions should be considered dynamic at runtime. It
+provides a balance between user control, flexibility, and ease of integration.
+Compared to the build flag it's more flexible, and compared to the environment
+variable it's less opaque.
+
+### Option 1: Using DNNL_RUNTIME_DIM_VAL
+
+With this option, the user will select one or more dimensions to have the
+DNNL_RUNTIME_DIM_VAL special value during primitive creation. oneDNN will then
+dispatch to the first implementation that supports the given set of runtime
+dimensions.
+<u>Pros:</u>
+- The most fine-grained approach
+- Ties into existing mechanisms
+- Can be toggled without a rebuild
+
+<u>Cons:</u>
+- Several primitives have no implementations that support runtime dims -
+cannot be adopted until such implementations have been developed
+- Support for all possible runtime dim configurations is difficult
+- Requires users to know which dimensions will have dynamic shapes ahead of
+time
+
+### Option 2: Build flag
+
+With this option, the build flag `ONEDNN_PREFER_REUSABLE_KERNELS` would be
+added to select the preferred functionality. Default is `off`, and reusable
+kernels will be chosen preferentially when set to `on`.
+<u>Pros:</u>
+- Easy integration
+
+<u>Cons:</u>
+- Requires a rebuild to change the option
+- The setting is not apparent in the code, leads to opaque behaviour
+- Requires internal changes to detect implementation reusability
+
+### Option 3: Environment variable
```

Comment:
I think the main question here is, do we want to provide a per-primitive control or a single knob for all primitives? The per-primitive option looks more flexible to me and the primitive attribute (option №4) would be the way to go. User could introduce their own environment variable that would control the primitive attribute if needed.